release/cc/create_artifacts.py

```diff
@@ -0,0 +1,184 @@
+          config_version=container_configs.version,
+          bazel_version=bazel_version)
+
+      with open(container_configs.get_target_build_path(), "a+") as build_file:
+        if target not in build_file.read():
+          tpl_file_path = os.path.join(GIT_ROOT, "release", "cc",
+                                       "%s.tpl" % config.config_type)
+          with open(tpl_file_path, "r") as tpl_file:
+            tpl = Template(tpl_file.read()).substitute(
+                DATE=get_date(),
+                DISTRO=container_configs.distro,
+                CONFIG_VERSION=container_configs.version,
+                BAZEL_VERSION=bazel_version,
+                NAME=container_configs.image,
+                SHA=sha)
+
+            build_file.write(tpl)
+
+
+def generate_toolchain_definition(container_configs_list, bazel_version):
+  """Generate new cpp toolchain definitions.
+
+  Example cpp toolchain definitions for clang-ubuntu container are located in
+  configs/ubuntu16_04_clang/1.0/bazel_0.15.0/cpp/.
+
+  There is one BUILD file to contain all cpp toolchain definitions for each
+  config type (e.g. default, msan) per container per Bazel version.
+
+  If the file already exist in this repo, the script will delete it and
+  generate new one.
+
+  Args:
+    container_configs_list: list of ContainerConfigs, the list of
+      ContainerConfigs to generate configs for.
+    bazel_version: string, the version of Bazel used to generate the configs.
+
+  Returns:
+    None
+  """
+
+  for container_configs in container_configs_list:
+
+    cpp_dir = os.path.dirname(container_configs.get_cpp_build_path())
+
+    # Remove old cpp directory if exists.
+    if os.path.isdir(cpp_dir):
+      shutil.rmtree(cpp_dir)
+    os.makedirs(cpp_dir)
+
+    with open(container_configs.get_cpp_build_path(), "w") as build_file:
+      # Write license header.
+      with open(LICENCE_TPL, "r") as license_header:
+        build_file.write(license_header.read())
+
+      # Write extra license string required for BUILD file.
+      build_file.write(BUILD_EXTRA_LICENCE)
+
+    for config in container_configs.configs:
+
+      with open(container_configs.get_cpp_build_path(), "a") as build_file:
+        with open(CPP_TPL, "r") as tpl_file:
+
+          # Merge constraint lists. Remove duplicates while perserving order.
+          constraints = container_configs.constraints
+          for constraint in config.constraints:
+            if constraint not in constraints:
+              constraints.append(constraint)
+
+          tpl = Template(tpl_file.read()).substitute(
+              TYPE=config.config_type,
+              CONFIG_VERSION=container_configs.version,
+              BAZEL_VERSION=bazel_version,
+              PACKAGE=container_configs.package,
+              EXTRA_CONSTRAINTS="\n".join(
+                  [("\"%s\"," % constraint) for constraint in constraints]))
+
+          build_file.write(tpl)
+
+      subprocess.call(
```

Comment:
why do you need to call buildifier on the generated BUILD files?

Reply:
To format the BUILD file, otherwise, the constraints won't be formatted nicely.